packages/minimongo/minimongo.js

```diff
@@ -782,8 +782,22 @@ LocalCollection.prototype.update = function (selector, mod, options, callback) {
   // generate an id for it.
   var insertedId;
   if (updateCount === 0 && options.upsert) {
-    var newDoc = LocalCollection._removeDollarOperators(selector);
+    var newDoc = {};
+     
+    if (LocalCollection._selectorIsId(selector)) {
+      selector = {_id: selector}
+    }
+
+    var selectorDocument = LocalCollection._removeDollarOperators(selector);
```

Comment:
Actually, we're not really building a selector at this point anymore, we're building a new upsert modifier. So maybe we could instead use a new variable called something like `upsertModifier`, to make our intent here more clear? So we could consider changing the code here to something like:

```js
...
let upsertModifier =
  LocalCollection._selectorIsId(selector) ? { _id: selector } : selector;
upsertModifier = LocalCollection._removeDollarOperators(upsertModifier);

const newDoc = {};
if (upsertModifier._id) {
  newDoc._id = upsertModifier._id;
  delete upsertModifier._id;
}

LocalCollection._modify(newDoc, {$set: upsertModifier});
LocalCollection._modify(newDoc, mod, {isInsert: true});
...
```